// ngx-lottie/bundles/ngx-lottie-server.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/platform-browser'), require('path'), require('ngx-lottie'), require('fs')) :
    typeof define === 'function' && define.amd ? define('ngx-lottie/server', ['exports', '@angular/core', '@angular/platform-browser', 'path', 'ngx-lottie', 'fs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['ngx-lottie-server'] = {}, global.ng.core, global.ng.platformBrowser, global.path, global['ngx-lottie'], global.fs));
}(this, (function (exports, core, platformBrowser, path, ngxLottie, fs) { 'use strict';

    function readFileWithAnimationData(path) {
        return new Promise(function (resolve, reject) {
            fs.readFile(path, function (error, data) {
                if (error) {
                    return reject(error);
                }
                resolve(data.toString());
            });
        });
    }

    function readAndTransferAnimationData(transferState, animations, pathsToAnimations) {
        var sources = [];
        var _loop_1 = function (i, length) {
            var path = pathsToAnimations[i];
            var source = readFileWithAnimationData(path).then(function (animationData) {
                transferAnimationData(transferState, animations[i], animationData);
            });
            sources.push(source);
        };
        for (var i = 0, length = animations.length; i < length; i++) {
            _loop_1(i, length);
        }
        return sources;
    }
    function transferAnimationData(state, animation, animationData) {
        animation = ngxLottie.transformAnimationFilenameToKey(animation);
        var key = platformBrowser.makeStateKey(animation);
        state.set(key, JSON.parse(animationData));
    }
    function appInitializerFactory(options, state) {
        var pathsToAnimations = resolveLottiePaths(options);
        var sources = readAndTransferAnimationData(state, options.preloadAnimations.animations, pathsToAnimations);
        return function () { return Promise.all(sources); };
    }
    function resolveLottiePaths(_a) {
        var preloadAnimations = _a.preloadAnimations;
        var folder = preloadAnimations.folder, animations = preloadAnimations.animations;
        var path$1 = path.join(process.cwd(), folder);
        return animations.map(function (animation) { return path.join(path$1, animation); });
    }

    var LOTTIE_SERVER_OPTIONS = new core.InjectionToken('LottieServerOptions');
    var LottieServerModule = /** @class */ (function () {
        function LottieServerModule() {
        }
        LottieServerModule.forRoot = function (options) {
            return {
                ngModule: LottieServerModule,
                providers: [
                    {
                        provide: LOTTIE_SERVER_OPTIONS,
                        useValue: options,
                    },
                    {
                        provide: core.APP_INITIALIZER,
                        useFactory: appInitializerFactory,
                        multi: true,
                        deps: [LOTTIE_SERVER_OPTIONS, platformBrowser.TransferState],
                    },
                ],
            };
        };
        return LottieServerModule;
    }());
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    LottieServerModule.decorators = [
        { type: core.NgModule }
    ];

    /**
     * Generated bundle index. Do not edit.
     */

    exports.LottieServerModule = LottieServerModule;
    exports.ɵa = LOTTIE_SERVER_OPTIONS;
    exports.ɵb = appInitializerFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
